### DISCUSSION TITLE
Investigate Table Partitioning

### DISCUSSION THREAD

**Discussion by janbjorge** [janbjorge/pgqueuer#84] [Ideas]
Evaluate the feasibility and benefits of implementing table partitioning in PgQueuer to improve performance and scalability.

**Objective**

Assess whether partitioning `queue_data` using PostgreSQL's `PARTITION BY LIST (...)` could enhance the system's efficiency and manageability:

- `queue_data_new` for new jobs
- `queue_data_processing` for active jobs
- `queue_data_standby` for standby jobs

**Expected Outcomes**

- Determine potential improvements in query performance and I/O operations.
- Evaluate scalability and maintenance benefits from partitioning.
- Identify challenges and considerations for integrating partitioning with existing features.

**Action Items**

- Research partitioning strategies and their alignment with PgQueuer’s architecture.
- Prototype potential partitioning schemas and measure performance impacts.
- Develop a clear migration path for existing installations.

**Reference**

- [PostgreSQL Partitioning](https://www.postgresql.org/docs/current/ddl-partitioning.html)
- [Queues in Postgres | Postgres.FM 042 | #PostgreSQL #Postgres podcast ](https://www.youtube.com/watch?v=mW5z5NYpGeA&t=1766s)

**Comment by beratdogan** [comment]
Dynamic partition management might not be feasible for some environments, like non-privileged application PG role. Therefore, it should be optional, and the `in tx; delete record from previous tables, add to next tables` add to  strategy could be useful in those cases.

Another thing to consider about partition management is locks. I remember some operations require `access exclusive` locks on the parent table. Locks should be considered on live environments.

**Reply by janbjorge** [reply]
Thanks for your insights on dynamic partition management and locking concerns. These points are critical and highlight the complexities of integrating partitioning in diverse environments.

Given the intricacies involved, I need to further investigate PostgreSQL’s partitioning and its impact on PgQueuer. I’ll continue to research and prototype before making any decisions.

I appreciate your input and will keep you updated on my progress.
